### PR TITLE
SI-9113: WrapAsScala can create exception-prone Wrappers

### PR DESCRIPTION
**Pull request by RomanHargrave**
Review by @axel22 @phaller .

Fixes issue SI-9113 by checking for and returning null values instead of passing them to a wrapper as an underlying collection, which can in turn lead to unexpected behaviour.

This is a follow-up to #4266, in which I neglected to squash commits.

### CODE REVIEW

**Comment by Ichoran** [comment]
Which, if any, of the existing methods fail fast and now will fail more slowly with a NPE later on?

**Comment by RomanHargrave** [comment]
@Ichoran This should not cause a delayed NPE anywhere. The problem detailed in SI-9113 is that if you are using the implicit converters (or decorators, as they use common logic) to wrap a java collection reference, and that reference is null, it will return a collection wrapper like those implemented in [Wrappers.scala](/scala/scala/blob/2.11.x/src/library/scala/collection/convert/Wrappers.scala). Since these wrappers will be erased to their destination type, it is not possible to access the underlying collection (not that there exists a clean method by which to accomplish this given you know the type _is_ a wrapper implementation) in order to determine whether or not it is a null reference. This creates a rather pernicious issue, as you will wind up with an object reference that is itself _not_ null, and that when accessed will throw an NPE inside of it self rather than at the actual caller (which, due to Scala->Java magic, can be several stack levels of difference). Due to the aforementioned traits, this will bypass any nullity checks performed in client code, and Option(_) will return Some[_](...) instead of None when passed the faulty wrapper. As such, in order to safely deal with an implicitly converted collection that might have been null before conversion, you must encase it in a try/catch block whenever you access or manipulate it. The unfortunate bit about this approach is that it can be hard to avoid catching exceptions that might possibly occur _inside_ the underlying collection (for instance: when you convert a custom collection that might throw that error), which could lead to some rather confusing debugging if the problem-causing exception is dismissed as regular functionality.

As for how it affects code that was written to accommodate this behaviour, nothing should be different. Currently, when you attempt to call a wrapper function, it will throw an NPE _inside_ the wrapper. After the fix, an NPE will still be throw but rather than _inside_ the wrapper, it will be _at_ the caller (or close to it). As such, any try/catch statements in place to handle this potential behaviour can be either left in place or replaced with Option/match handling or a nullity check.

**_In short**_, none of the implicit converters will fail (in the sense of rasing an exception) any differently from how they could have before the proposed modifications. Furthermore, any code you have in place to handle NPEs thrown when accessing null collections will still work.

That being said, I would suggest that you test your software with the modified library in place, as you would with any other update.

**Comment by Ichoran** [comment]
Thanks for the lengthy response, but I think you misunderstood my concern.

If any existing wrapper touches the wrapped collection in the initializer (e.g. there's  lazy val somewhere), it could fail fast.  So I just wanted, e.g., "No wrapper accesses the wrapped collection in the initializer", or "some collection does, so it will fail fast on construction".

I read through the code myself and did not see any instances where the initializer touches a possibly null collection in any of the wrappers, so I believe that the code will fail at the same locations as before (with a little less on the stack).

There is a secondary issue: behavior of some data-aware constructs will change.  E.g. Option(x.asScala) will be None instead of Some(y) where y will throw a NPE.  I'd argue that this is what one wants, but we should keep our eye on it since it's targeted to a point reliease.

**Comment by RomanHargrave** [comment]
I can second you on the wrapper classes not accessing the underlying collection during initialization.

**Comment by RomanHargrave** [comment]
Blasted mobile browsers these days...

**Comment by Ichoran** [comment]
LGTM but I would like a second opinion from someone on whether the behavior change on Option(javaCollThatIsNull.asScala) is fine for a point release.  I would like to say yes: the whole point of putting it in an option is to catch this sort of thing!

**Comment by axel22** [comment]
LGTM too, but to be safe, I would add it in the next major release.

Would be tempted to put the `migration` annotation as well, but that might be too obtrusive.

**Comment by retronym** [comment]
I think we should strive for a consistent policy with other implicits in the library.

```
scala> (null: String): Seq[Char]
res0: Seq[Char] = null

scala> (null: java.math.BigDecimal): scala.BigDecimal // throws since 2.11
java.lang.IllegalArgumentException: null value for BigDecimal
  at scala.math.BigDecimal.<init>(BigDecimal.scala:406)
  at scala.math.BigDecimal$.apply(BigDecimal.scala:334)
  at scala.math.BigDecimal$.apply(BigDecimal.scala:331)
  at scala.math.BigDecimal$.javaBigDecimal2bigDecimal(BigDecimal.scala:346)
  ... 33 elided

scala> (null: Array[Any]): Seq[Any]
res3: Seq[Any] = null
```

I remember some discussion of this sort of problem before but I can't seem to find it. Will keep looking.

**Comment by axel22** [comment]
True. It is not likely that every implicit writer will be careful enough to check for `null`ness. The current behavior could be more consistent with the Scala ecosystem in general, not just the standard library.

**Comment by som-snytt** [comment]
My comment on the previous PR was that `(null: java.util.Iterator[Int]).asScala` should throw because it's a selection. It makes me wonder why the compiler doesn't enforce it at runtime.

```
scala> implicit class fer[A](a: A) { def f = 42 }
defined class fer

scala> null.f
<console>:12: error: value f is not a member of Null
              null.f
                   ^

scala> (null: Any).f
res4: Int = 42
```

Enforcement could be injected in the `implicit def` of implicit classes.

**Comment by adriaanm** [comment]
Since this makes (potential) NPE crashes easier to understand without materially changing the behavior, I think we can merge this into 2.11.x.

I feel bad asking for a commit message rewrite after so much time, so I've reworded and resubmitted. See linked PR.

**Comment by RomanHargrave** [comment]
@retronym Just had a thought at random this morning. Wouldn't it make more since for the decorated converters (.asJava, .asScala) to throw exceptions, but for the implicit converters to return null? My logic is that it would make debugging simpler. For instance, when passing an X collection to a function that wants a Y collection with implicit conversion, it will appear in source form as if that reference is being passed to said function (when there is really an intermediate function call inserted during compilation), and as such, it would make more sense for that function to throw the exception from within. On the other hand, when using the decorative method, it appears as if you are calling a function provided by X collection to get a Y collection. In the latter (decorator) case, it would make more sense to throw an exception from inside the conversion method, as it would be clear that the programmer _explicitly_ wanted to call a member function of X collection, rather than pass the reference to X to the method (whether it be null or not). What do you think?

**Comment by axel22** [comment]
@som-snytt if anybody wanted to write e.g. a method that runs a `null` into an `Option`, this would be much harder if compiler were to throw an exception. I guess a hypothetical universal `NonNullAnyVal` trait might be a better way to achieve auto-`NullPointerException` conversions.
